Keen.NetStandard/Dataset/Datasets.cs

```diff
@@ -0,0 +1,306 @@
+            }
+
+            return JsonConvert.DeserializeObject<DatasetDefinitionCollection>(responseString,
+                                                                              SERIALIZER_SETTINGS);
+        }
+
+        public async Task<IEnumerable<DatasetDefinition>> ListAllDefinitionsAsync()
+        {
+            var allDefinitions = new List<DatasetDefinition>();
+            var firstSet = await ListDefinitionsAsync(MAX_DATASET_DEFINITION_LIST_LIMIT)
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            if (null == firstSet?.Datasets)
+            {
+                throw new KeenException("Failed to fetch definition list");
+            }
+
+            if (!firstSet.Datasets.Any())
+            {
+                return allDefinitions;
+            }
+
+            if (firstSet.Count <= firstSet.Datasets.Count())
+            {
+                return firstSet.Datasets;
+            }
+
+            allDefinitions.AddRange(firstSet.Datasets);
+
+            do
+            {
+                var nextSet = await ListDefinitionsAsync(MAX_DATASET_DEFINITION_LIST_LIMIT,
+                                                         allDefinitions.Last().DatasetName)
+                    .ConfigureAwait(continueOnCapturedContext: false);
+
+                if (null == nextSet?.Datasets || !nextSet.Datasets.Any())
+                {
+                    throw new KeenException("Failed to fetch definition list");
+                }
+
+                allDefinitions.AddRange(nextSet.Datasets);
+            } while (firstSet.Count > allDefinitions.Count);
+
+            return allDefinitions;
+        }
+
+        public async Task DeleteDatasetAsync(string datasetName)
+        {
+            if (string.IsNullOrWhiteSpace(datasetName))
+            {
+                throw new KeenException("A dataset name is required.");
+            }
+
+            if (string.IsNullOrWhiteSpace(_masterKey))
+            {
+                throw new KeenException("An API masterkey is required to get dataset results.");
+            }
+
+            var responseMsg = await _keenHttpClient
+                .DeleteAsync(GetDatasetUrl(datasetName), _masterKey)
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            var responseString = await responseMsg
+                .Content
+                .ReadAsStringAsync()
+                .ConfigureAwait(continueOnCapturedContext: false);
+
+            if (HttpStatusCode.NoContent == responseMsg.StatusCode)
+            {
+                return;
+            }
+
+            var response = JObject.Parse(responseString);
+
+            KeenUtil.CheckApiErrorCode(response);
+
+            throw new KeenException($"Request failed with status: {responseMsg.StatusCode}");
+        }
+
+        public async Task<DatasetDefinition> CreateDatasetAsync(DatasetDefinition dataset)
+        {
+            if (string.IsNullOrWhiteSpace(_masterKey))
```

Comment:
Part of it is because yeah, unfortunately it's created even if not used (ugh...), a lot of what's in Issues #55 and #56 comes out of PR #51, but also, getting a Dataset definitions or results only requires a read key.

Reply:
I'd considered creating an `Attribute` to describe key-ness for these APIs the way we use decorators in the Python SDK, but haven't gotten around to looking at how well that'd work.

Reply:
I noticed that while ReadKey is accepted by the back end for getting results/defs, our `Datasets` impl was demanding MasterKey in those cases, so I fixed that locally, and it'll be part of the final version of this PR.

Reply:
I'm curious how that might work. Can `Attribute`s work like function decorators/closures, or does some calling code have to check for the attribute and then do something?

Reply:
They aren't, I think, as powerful as a Python decorator, but more like Java annotations, so there'd be runtime-emitted info in the metadata that would allow some wrapper code to reflect on the method in question and decide if the proper keys are set, and how/whether to proceed. We can't fail compile like we could with a full-on metaprogramming solution in C++, but it'd be more declarative and allow us to share that key-checking logic across classes that have no common ancestor in the inheritance hierarchy.